# Guided Capstone Step 4. Pre-Processing and Training Data Development

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
#load python packages
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv('step3_output.csv')
df.describe()
df.head()

Name       state  ...  NightSkiing_ac  clusters
0                 Alyeska Resort      Alaska  ...           550.0         1
1            Eaglecrest Ski Area      Alaska  ...             0.0         1
2               Hilltop Ski Area      Alaska  ...            30.0         1
3               Arizona Snowbowl     Arizona  ...             0.0         2
4  Yosemite Ski & Snowboard Area  California  ...             0.0         0

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)

In [5]:
df.head()

Name  summit_elev  ...  Wisconsin  Wyoming
0                 Alyeska Resort         3939  ...          0        0
1            Eaglecrest Ski Area         2600  ...          0        0
2               Hilltop Ski Area         2090  ...          0        0
3               Arizona Snowbowl        11500  ...          0        0
4  Yosemite Ski & Snowboard Area         7800  ...          0        0

[5 rows x 54 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**


In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

In [7]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name', 'AdultWeekend'], axis=1)

In [8]:
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

In [9]:
pip install scikit-learn

In [10]:
from sklearn.preprocessing import StandardScaler

In [11]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

In [12]:
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [13]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [14]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [15]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [18]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
# You can plug y_test and y_pred into the functions to evaluate the model
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.8666718270947033
6.057373868531214


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [19]:
print(lm.intercept_)

70.3031250264137


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [20]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)


Coefficient
AdultWeekday       2.512814e+01
North Carolina     3.789821e+00
Wisconsin          3.669928e+00
averageSnowfall    3.650496e+00
Runs               2.159494e+00
Pennsylvania       2.115235e+00
Colorado           1.576719e+00
Oregon             1.550125e+00
fastQuads          1.456539e+00
projectedDaysOpen  1.394526e+00
vertical_drop      1.355265e+00
Wyoming            1.336106e+00
SkiableTerrain_ac  1.149478e+00
Montana            1.118729e+00
TerrainParks       1.022372e+00
Utah               9.963505e-01
clusters           9.858138e-01
fastSixes          9.495636e-01
New Hampshire      9.393705e-01
Maine              9.082541e-01
Washington         8.885873e-01
NightSkiing_ac     8.012327e-01
California         7.782238e-01
Michigan           7.667865e-01
New York           7.131844e-01
New Mexico         6.607198e-01
Vermont            6.597064e-01
Idaho              6.385453e-01
daysOpenLastYear   5.996126e-01
base_elev          5.891600e-01
total_chairs       5.715638e-01
Arizona            5.606757e-01
summit_elev        5.113570e-01
Snow Making_ac     5.059533e-01
surface            3.867508e-01
South Dakota       3.625574e-01
Iowa               3.040301e-01
fastEight          3.019777e-01
Massachusetts      3.015233e-01
Nevada             2.937742e-01
Minnesota          2.812465e-01
triple             2.436589e-01
double             2.096471e-01
Connecticut        1.861940e-01
Indiana            1.771417e-01
trams              1.624737e-01
Missouri           1.452045e-01
quad               1.368717e-01
Alaska             1.194423e-01
yearsOpen          1.188581e-01
LongestRun_mi      3.156922e-02
Ohio               8.881784e-16

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [21]:
#remove states from the columns we are working with 
df.columns[1:-28]

Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [22]:
no_state_cols = list(df.columns[1:-28])
no_state_cols.remove('AdultWeekend')
print(no_state_cols)

['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters']


In [23]:
X_scaled = scaler.fit_transform(df.loc[:,no_state_cols])
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.9291280313240549
5.777338132110073


In [24]:
pd.DataFrame(abs(lm.coef_), no_state_cols, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)

Coefficient
summit_elev          31.152149
base_elev            26.733777
AdultWeekday         24.809011
vertical_drop         6.976485
Runs                  3.463302
fastQuads             2.831948
daysOpenLastYear      2.006973
surface               1.364849
fastEight             1.309319
fastSixes             1.204550
averageSnowfall       1.183657
projectedDaysOpen     1.113475
total_chairs          0.854646
quad                  0.766359
triple                0.688056
TerrainParks          0.686887
LongestRun_mi         0.615522
NightSkiing_ac        0.397402
trams                 0.324140
Snow Making_ac        0.203164
double                0.179379
yearsOpen             0.152957
clusters              0.111112
SkiableTerrain_ac     0.066581

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [25]:
#remove summit elevation and base elevation
no_elev_cols = no_state_cols
no_elev_cols.remove('summit_elev')
no_elev_cols.remove('base_elev')
print(no_elev_cols)

['vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters']


In [26]:
X_scaled = scaler.fit_transform(df.loc[:,no_state_cols])
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.9308202013748784
5.5690074111119765


In [27]:
pd.DataFrame(abs(lm.coef_), no_state_cols, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday         24.906110
fastQuads             2.570505
Runs                  2.280246
daysOpenLastYear      1.836427
projectedDaysOpen     1.603119
surface               1.193095
fastSixes             1.064718
SkiableTerrain_ac     1.037847
NightSkiing_ac        0.967626
fastEight             0.878741
vertical_drop         0.874390
averageSnowfall       0.858519
total_chairs          0.828908
LongestRun_mi         0.811538
quad                  0.767938
triple                0.704772
TerrainParks          0.605126
Snow Making_ac        0.299165
trams                 0.221898
yearsOpen             0.154867
double                0.143702
clusters              0.097067

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | .867 | 6.057 |-|
| Model 2. | .929 | 5.777 |'state'|
| Model 3. | .931 | 5.569 |'state','summit_elev','base_elev'|

Model Selection:

Here, we will choose Model 3 as it eliminates Big Mountain Ski Resort's attributes that we cannot change, ie. state(location), summit elevation, and base elevation. Furthermore, Model 3's explained variance is the highest at .931 and the mean squared error is the lowest at 5.569.